### ISSUE TITLE
question about COCO detection

### ISSUE THREAD

**Issue by touristourist** [Haochen-Wang409/HPM#2]
Hi, nice to read such an interesting paper. The pretraining code is well-detailed,  but this repository lacks information regarding COCO detection downstream finetuning. I am curious about how to conduct COCO detection finetuning after MAE pretraining. Could you provide more details about the detection architecture implementation and experiment config?

**Comment by Haochen-Wang409** [comment]
Hi, thanks a lot for your attention!
To evaluate our HPM on COCO, we simply follow the configuration of [ViTDet](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet), and the detailed configuration is [here](https://github.com/facebookresearch/detectron2/blob/main/projects/ViTDet/configs/COCO/mask_rcnn_vitdet_b_100ep.py).

Two modifications are needed:
- ```train.init_checkpoint``` is modified to our own pre-trained model.
- Setting ```train.max_iter=88750```. This is because with a batch size of 16, ```1x = 88750 iters * 16 images/iter / 1420000 images/schedule```.

**Comment by touristourist** [comment]
> 

Thank you so much for your quick reply! And I have some further questions:
1.  When I use detectron2 to load HPM pre-trained model, I encounter key mismatches. In the ViTDet, the keys have `backbone.net.` prefix, while in HPM, the prefix is `module.`. Even after aligning their prefixes, there are still some mismatched keys between HPM-encoder and ViTDet-backbone, e.g., {`cls_token`, `decoder_pos_embed`, `mask_token`} exist only in HPM, while `backbone.net.blocks.0.attn.{rel_pos_h, rel_pos_w`} exists only in ViTDet. It appears that ViTDet uses relative position while HPM does not. So I wonder how you resolved these mismatched keys? Do they have a bad impact?
2.  Regarding training iterations, if I can only support a batch size of 8, should I double the `train.max_iter`? Will the performance degrade with the smaller batch size?
3.  I have tried HPM-pretrained and random-init ViTDet on COCO simply with my own hyper-parameter. But in the early finetuning stage, the HPM-pretrained one has a lower mAP than the random-init one. Is this normal?

Looking forward to your reply! Thanks!

**Comment by Haochen-Wang409** [comment]
1. Simply aligning their prefixes is all you needed. Other pre-trained models, e.g., MAE, do not utilize relative position encodings at the same time.
2. Yes, it is supposed to double the ```train.max_iter```. As for performance degradation with smaller a batch size, I am afraid I cannot answer you because we didn't conduct experiments with a batch size of 8.
3. Actually, if you pre-trained HPM with only 200 epochs, it might be possible. It is better to reproduce the performance of MAE using their official weights first to make sure that the fine-tuning is going right.

**Comment by touristourist** [comment]
> 1. Simply aligning their prefixes is all you needed. Other pre-trained models, e.g., MAE, do not utilize relative position encodings at the same time.
> 2. Yes, it is supposed to double the `train.max_iter`. As for performance degradation with smaller a batch size, I am afraid I cannot answer you because we didn't conduct experiments with a batch size of 8.
> 3. Actually, if you pre-trained HPM with only 200 epochs, it might be possible. It is better to reproduce the performance of MAE using their official weights first to make sure that the fine-tuning is going right.

OK, thank you! I will try it again.

**Comment by Haochen-Wang409** [comment]
I hope you can successfully reproduce the results :)

**Comment by touristourist** [comment]
Hi, Haochen. A tiny question again. About the lr scheduler, if `train.max_iter` is set to 88750 in the [config](https://github.com/facebookresearch/detectron2/blob/main/projects/ViTDet/configs/COCO/mask_rcnn_vitdet_b_100ep.py),  how did you set the value of milestones below, which controls when to conduct stepLRScheduler. 
```
lr_multiplier = L(WarmupParamScheduler)(
    scheduler=L(MultiStepParamScheduler)(
        values=[1.0, 0.1, 0.01],
        milestones=[163889, 177546],
        num_updates=train.max_iter,
    ),
    warmup_length=250 / train.max_iter,
    warmup_factor=0.001,
)
```

**Comment by Haochen-Wang409** [comment]
I am sorry that I failed to fully explain our modifications.
As the original ```train.max_iter=184375```, we simply compute the new milestones when ```train.max_iter=88750``` by ```int(163889/184375) * 88750 = 78889``` and ```int(177546/184375) * 88750 = 85463```.

**Comment by Haochen-Wang409** [comment]
Also, you are supposed to change the ```dataloader.train.total_batch_size=16``` in ```configs/common/coco_loader_lsj.py```.

**Comment by touristourist** [comment]
OK, thank you for your careful answer!

**Comment by touristourist** [comment]
Hi, I‘m sorry I can not reproduce the performance of the paper. In Table 5 of the paper, the bbox mAP of 200-epochs pretrained MAE model on COCO is 40.45. However, I pretrained my own 200-epochs MAE model with imagenet-top1 ACC 82.31, and then I fine-tuned it on COCO with 36 epochs. The bbox mAP is 45.9475 as below.  
![screencut](https://github.com/Haochen-Wang409/HPM/assets/28351316/4a89be31-ca94-4d35-889e-62b0a5036500)
And I downloaded and finetuned the HPM 800-epoch model in this repo, the result of bbox mAP on COCO is 46.1888, which is also much higher than that in the paper.

Here is my config of 200-epochs pretrained MAE. I wonder whether there is something wrong I omit.
```
from functools import partial
from fvcore.common.param_scheduler import MultiStepParamScheduler

from detectron2 import model_zoo
from detectron2.config import LazyCall as L
from detectron2.solver import WarmupParamScheduler
from detectron2.modeling.backbone.vit import get_vit_lr_decay_rate

from ..common.coco_loader_lsj import dataloader


model = model_zoo.get_config("common/models/mask_rcnn_vitdet.py").model

# Initialization and trainer settings
train = model_zoo.get_config("common/train.py").train
train.amp.enabled = True
train.ddp.fp16_compression = True

# train.init_checkpoint = (
#     "detectron2://ImageNetPretrained/MAE/mae_pretrain_vit_base.pth?matching_heuristics=True"
# )

train.init_checkpoint = (
    "./pretrained_models/mae/mae_82.31/checkpoint-99process.pth"
)

train.checkpointer.period=10000

train.output_dir = "./output/mask_rcnn_vitdet_b_100ep_mae_baseline_82.31/"

dataloader.train.total_batch_size = 8


# Schedule
train.max_iter = 177500   # hpm for batch size 8


lr_multiplier = L(WarmupParamScheduler)(
    scheduler=L(MultiStepParamScheduler)(
        values=[1.0, 0.1, 0.01],
        milestones=[157778, 170926],
        num_updates=train.max_iter,
    ),
    warmup_length=250 / train.max_iter,
    warmup_factor=0.001,
)

# Optimizer
optimizer = model_zoo.get_config("common/optim.py").AdamW
optimizer.params.lr_factor_func = partial(get_vit_lr_decay_rate, num_layers=12, lr_decay_rate=0.7)
optimizer.params.overrides = {"pos_embed": {"weight_decay": 0.0}}

```

**Comment by Haochen-Wang409** [comment]
All experiments on COCO are trained with only 1x schedule, i.e., 12 epochs.

**Comment by touristourist** [comment]
> All experiments on COCO are trained with only 1x schedule, i.e., 12 epochs.

Thanks, I'd like to make sure again that which model did you use as initialization in COCO exp? the pretrained model or the IN1K-finetuned model?

**Comment by Haochen-Wang409** [comment]
The pre-trained model.

**Comment by Haochen-Wang409** [comment]
The pre-trained model.
